pkg/scheduler/apis/config/v1alpha1/defaults.go

```diff
@@ -55,22 +58,37 @@ func SetDefaults_KubeSchedulerConfiguration(obj *kubeschedulerconfigv1alpha1.Kub
 		}
 	}
 
-	if host, port, err := net.SplitHostPort(obj.HealthzBindAddress); err == nil {
+	defaultBindAddress := net.JoinHostPort("0.0.0.0", strconv.Itoa(ports.InsecureSchedulerPort))
+	if obj.HealthzBindAddress == nil {
+		obj.HealthzBindAddress = &defaultBindAddress
+	}
+	if host, port, err := net.SplitHostPort(*obj.HealthzBindAddress); err == nil {
 		if len(host) == 0 {
 			host = "0.0.0.0"
 		}
-		obj.HealthzBindAddress = net.JoinHostPort(host, port)
-	} else {
-		obj.HealthzBindAddress = net.JoinHostPort("0.0.0.0", strconv.Itoa(ports.InsecureSchedulerPort))
```

Comment:
this case got dropped (existing value is non-nil, but does not include port)

Reply:
Now that you mention it, it looks like this didn't quite cover that case originally either, since it defaults the address to `0.0.0.0` even if it's just missing the port, so I'll fix these up to use the address that's set and default to `0.0.0.0` only if that's empty

Reply:
In my changes though, I did cover these `else` cases by declaring the `defaultBindAddress` above both, and setting it before the error checks (as opposed to after)

Reply:
> In my changes though, I did cover these `else` cases by declaring the `defaultBindAddress` above both (https://github.com/kubernetes/kubernetes/pull/83619/files/b019c2ec6ba787ac1c24ce291fb8490420d951af#diff-9d72b6493a5482d8a2829af1390c5733R61), and setting it before the error checks (as opposed to after)

I don't think that's accurate. You only set it in the nil case, when the value is completely absent from the config. When just the address (`0.0.0.0`) or just the port (`:1234`) is set, the nil path would not be exercised.

Reply:
Actually, it seems that `net.SplitHostPort` doesn't even return the host if the port is missing ([playground example](https://play.golang.org/p/j-F_PGIoggO)), so it becomes more difficult here to properly implement the case you described. If the existing value is `1.2.3.4` (no port), then `SplitHostPort` only returns an error `missing port in address`. We could do parsing (regex?) on our own to determine if the original value is a valid host, then append the port, while we would also have to string-compare the error message itself, while making sure there are no other errors in it. @liggitt what do you think?

Reply:
would this work: if splithostport returns an err, assume the value does not have a port, call call net.ParseIP on the value to make sure it is a valid bind address, and if that returns non-nil, append the default port?

Reply:
@liggitt you are right, my original changes would only work in the nil case. I've added the suggestion you make the check the output from net.ParseIP, and I think the combination of these has enabled us to check each edge case. For confirmation, I refactored the unit test for these defaults to use table tests and added these cases. Please take another look, and thank you for your feedback so far

Reply:
(changes pushed to new commit)